src/index.ts

```diff
@@ -771,20 +775,96 @@ class Spanner extends GrpcService {
         callback(err);
         return;
       }
-      const clientName = config.client;
+      this.getGaxClient_(config, (err, gaxClient) => {
+        if (err) {
+          callback(err);
+          return;
+        }
+        let reqOpts = extend(true, {}, config.reqOpts);
+        reqOpts = replaceProjectIdToken(reqOpts, projectId!);
+        const requestFn = gaxClient[config.method].bind(
+          gaxClient,
+          reqOpts,
+          config.gaxOpts
+        );
+        callback(null, requestFn);
+      });
+    });
+  }
+
+  /**
+   * get GAX client. This will retrive the end point uris for specific instance and cache the client accordingly.
+   *
+   * @private
+   *
+   * @param {string} projectId Request config
+   * @param {object} config Request config
+   * @param {function} callback Callback function
+   */
+  getGaxClient_(config, callback) {
+    let clientName = config.client;
+    if (
+      clientName !== 'SpannerClient' ||
+      !this.options.enableResourceBasedRouting
+    ) {
       if (!this.clients_.has(clientName)) {
-        this.clients_.set(clientName, new gapic.v1[clientName](this.options));
+        this.setSpannerClient(clientName, config, this.options);
       }
       const gaxClient = this.clients_.get(clientName)!;
-      let reqOpts = extend(true, {}, config.reqOpts);
-      reqOpts = replaceProjectIdToken(reqOpts, projectId!);
-      const requestFn = gaxClient[config.method].bind(
-        gaxClient,
-        reqOpts,
-        config.gaxOpts
+      callback(null, gaxClient);
+      return;
+    }
+
+    if (!config.instanceId) {
+      const error = new Error('instanceId is required.');
+      callback(error);
+      return;
+    }
+    const instanceId = config.instanceId;
+    clientName = `${clientName}-${instanceId}`;
+
+    if (!this.clients_.has(clientName)) {
+      this.instances_.get(instanceId)!.getMetadata(
+        {
+          fieldNames: ['endpointUris'],
+        },
+        (err, instance) => {
+          if (err) {
+            if (err.code === 7) {
```

Comment:
I'm a bit worried about checking for code 7 since it's not clear what this is. If it's possible to be more explicit about checking for the PERMISSION_DENIED error, that would be great. If not, add a comment to say that this is what's being checked here.

Reply:
I was looking into `enum` but it's not there for `status`, just created `const` variable name with `PERMISSION_DENIED`.

Reply:
Done!!